core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala

```diff
@@ -0,0 +1,207 @@
+        }
+      info(s"Processed offset for epoch request for epoch:$requestedEpoch and returning offset $offset from epoch list of size ${epochs.size}")
+      offset
+    }
+  }
+
+  /**
+    * Removes all epoch entries from the store with start offsets greater than or equal to the passed offset.
+    *
+    * @param offset
+    */
+  override def clearLatest(offset: Long): Unit = {
+    inWriteLock(lock) {
+      val before = epochs
+      if (offset >= 0 && offset <= latestOffset()) {
+        epochs = epochs.filter(entry => entry.startOffset < offset)
+        flush()
+        info(s"Cleared latest ${before.toSet.filterNot(epochs.toSet)} entries from epoch cache based on passed offset $offset leaving ${epochs.size} in EpochFile for partition $topicPartition")
+      }
+    }
+  }
+
+  /**
+    * Clears old epoch entries. This method searches for the oldest epoch < offset, updates the saved epoch offset to
+    * be offset, then clears any previous epoch entries.
+    *
+    * @param offset the offset to clear up to
+    */
+  override def clearEarliest(offset: Long): Unit = {
+    //TODO - Ben to double check this - committing now as about to get on plane.
+    inWriteLock(lock) {
+      val before = epochs
+      if (offset >= 0 && earliestOffset() <= offset) {
+        val earliest = epochs.filter(entry => entry.startOffset <= offset)
+        if(earliest.size > 0) {
+          epochs = epochs --= earliest
+          //If the offset is less than the earliest offset remaining, add previous epoch back, but with an updated offset
+          if(offset < earliestOffset() || epochs.isEmpty)
+            new EpochEntry(earliest.last.epoch, offset) +=: epochs
+          flush()
+          info(s"Cleared earliest ${before.toSet.filterNot(epochs.toSet).size} entries from epoch cache based on passed offset $offset leaving ${epochs.size} in EpochFile for partition $topicPartition")
+        }
+      }
+    }
+  }
+
+  /**
+    * Delete all entries.
+    */
+  override def clear() = {
+    inWriteLock(lock) {
+      epochs.clear()
+      flush()
+    }
+  }
+
+  def epochEntries(): ListBuffer[EpochEntry] = {
+    epochs
+  }
+
+  private def earliestOffset(): Long = {
+    if (epochs.isEmpty) -1 else epochs.head.startOffset
+  }
+
+  private def latestOffset(): Long = {
+    if (epochs.isEmpty) -1 else epochs.last.startOffset
+  }
+
+  private def flush(): Unit = {
+    checkpoint.write(epochs)
+  }
+
+  def epochChangeMsg(epoch: Int, offset: Long) = s"NewEpoch->Offset{epoch:$epoch, offset:$offset}, ExistingEpoch->Offset:{epoch:$latestEpoch, offset$latestOffset} for Partition: $topicPartition"
+
+  def maybeWarn(epoch: Int, offset: Long) = {
+    if(epoch < latestEpoch())
+      warn(s"Received an PartitionLeaderEpoch assignment for an offset < LogEndOffset. " +
+        s"This implies messages have arrived out of order. ${epochChangeMsg(epoch, offset)}")
+    else if(epoch < 0 && epoch != RecordBatch.UNKNOWN_PARTITION_LEADER_EPOCH)
```

Comment:
It seems that we only append messages of format V2 and newer. So epoch is never expected to be <0 ?